File: go/15_3Sum.go
package main

import "sort"

func threeSum(nums []int) [][]int {
    ret := [][]int{}
    sort.Ints(nums)
    for left := 0; left < len(nums)-2; left++ {
        mid := left + 1
        right := len(nums) - 1
        if left > 0 && nums[left] == nums[left-1] {
            continue
        }
        for mid < right {
            total := nums[left] + nums[mid] + nums[right]
            if total == 0 {
                ret = append(ret, []int{nums[left], nums[mid], nums[right]})
                for mid < right && nums[mid] == nums[mid+1] {
                    mid++
                }
                for mid < right && nums[right] == nums[right-1] {
                    right--
                }
                mid++
                right--
            } else if total > 0 {
                right--
            } else {
                mid++
            }
        }
    }
    return ret
}

File: python/15_3Sum.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ret = []
        nums.sort()
        for left in range(len(nums) - 1):
            mid = left + 1
            right = len(nums) - 1
            if left > 0 and nums[left] == nums[left - 1]:
                continue
            while mid < right:
                total = nums[left] + nums[mid] + nums[right]
                if total == 0:
                    ret.append([nums[left], nums[mid], nums[right]])
                    while mid < right and nums[mid] == nums[mid + 1]:
                        mid += 1
                    while mid < right and nums[right] == nums[right - 1]:
                        right -= 1

                    mid += 1
                    right -= 1
                elif total > 0:
                    right -= 1
                else:
                    mid += 1

        return ret
